File: src/lib/stxm/regions.ts
import "server-only";

/**
 * Region geometry helpers for line-scan sample/izero bar placement.
 */

function convolveSame(values: number[], kernel: number[]): number[] {
  const n = values.length;
  const kLen = kernel.length;
  const half = Math.floor(kLen / 2);
  const out = new Array<number>(n);
  for (let i = 0; i < n; i += 1) {
    let sum = 0;
    for (let j = 0; j < kLen; j += 1) {
      const idx = i + j - half;
      if (idx >= 0 && idx < n) {
        sum += (values[idx] ?? 0) * (kernel[j] ?? 0);
      }
    }
    out[i] = sum;
  }
  return out;
}

function kMeans1D(data: number[], k: number, randomState: number, maxIter = 100): number[] {
  const n = data.length;
  if (n === 0) {
    return [];
  }
  const sorted = [...data].sort((a, b) => a - b);
  const centroids: number[] = [];
  for (let i = 0; i < k; i += 1) {
    const idx = Math.min(n - 1, Math.floor(((i + 0.5) * n) / k));
    centroids.push(sorted[idx] ?? sorted[0] ?? 0);
  }
  const labels = new Array<number>(n).fill(0);
  let seed = randomState;
  const rand = (): number => {
    seed = (seed * 1664525 + 1013904223) >>> 0;
    return seed / 0x1_0000_0000;
  };
  for (let iter = 0; iter < maxIter; iter += 1) {
    let changed = false;
    for (let i = 0; i < n; i += 1) {
      let best = 0;
      let bestDist = Infinity;
      for (let c = 0; c < k; c += 1) {
        const dist = Math.abs((data[i] ?? 0) - (centroids[c] ?? 0));
        if (dist < bestDist) {
          bestDist = dist;
          best = c;
        }
      }
      if (labels[i] !== best) {
        labels[i] = best;
        changed = true;
      }
    }
    const newCentroids = new Array<number>(k).fill(0);
    const counts = new Array<number>(k).fill(0);
    for (let i = 0; i < n; i += 1) {
      const label = labels[i] ?? 0;
      newCentroids[label] = (newCentroids[label] ?? 0) + (data[i] ?? 0);
      counts[label] = (counts[label] ?? 0) + 1;
    }
    for (let c = 0; c < k; c += 1) {
      if ((counts[c] ?? 0) > 0) {
        centroids[c] = (newCentroids[c] ?? 0) / (counts[c] ?? 1);
      } else {
        centroids[c] = sorted[Math.floor(rand() * n)] ?? 0;
      }
    }
    if (!changed) {
      break;
    }
  }
  return labels;
}

function segmentSpatialRegions(
  image: number[][],
  nRegions = 3,
  profileColumns?: number,
  randomState = 0,
): { rowLabels: number[]; labelNames: string[] } {
  const nRows = image.length;
  const nCols = image[0]?.length ?? 0;
  const cols = profileColumns ?? Math.min(20, nCols);
  const profile: number[] = [];
  for (let row = 0; row < nRows; row += 1) {
    let sum = 0;
    for (let col = Math.max(0, nCols - cols); col < nCols; col += 1) {
      sum += image[row]?.[col] ?? 0;
    }
    profile.push(sum / cols);
  }
  const rowLabels = new Array<number>(nRows).fill(0);
  const raw = kMeans1D(profile, 3, randomState);
  const extents: number[] = [0, 0, 0];
  const meansIntensity: number[] = [0, 0, 0];
  const meanRow: number[] = [0, 0, 0];
  for (let idx = 0; idx < 3; idx += 1) {
    let count = 0;
    let intensitySum = 0;
    let rowSum = 0;
    for (let row = 0; row < nRows; row += 1) {
      if ((raw[row] ?? 0) === idx) {
        count += 1;
        rowSum += row;
        for (let col = 0; col < nCols; col += 1) {
          intensitySum += image[row]?.[col] ?? 0;
        }
      }
    }
    extents[idx] = count;
    meansIntensity[idx] = count > 0 ? intensitySum / (count * nCols) : 0;
    meanRow[idx] = count > 0 ? rowSum / count : 0;
  }
  let edgeIdx = 0;
  for (let idx = 1; idx < 3; idx += 1) {
    if ((extents[idx] ?? 0) < (extents[edgeIdx] ?? 0)) {
      edgeIdx = idx;
    }
  }
  const other = [0, 1, 2].filter((i) => i !== edgeIdx);
  const leftIdx = (meanRow[other[0] ?? 0] ?? 0) < (meanRow[other[1] ?? 0] ?? 0) ? other[0]! : other[1]!;
  const rightIdx = leftIdx === other[0] ? other[1]! : other[0]!;
  const sampleIdx =
    (meansIntensity[leftIdx] ?? 0) < (meansIntensity[rightIdx] ?? 0) ? leftIdx : rightIdx;
  const izeroIdx = sampleIdx === leftIdx ? rightIdx : leftIdx;
  for (let row = 0; row < nRows; row += 1) {
    const label = raw[row] ?? 0;
    if (label === sampleIdx) {
      rowLabels[row] = 0;
    } else if (label === edgeIdx) {
      rowLabels[row] = 1;
    } else if (label === izeroIdx) {
      rowLabels[row] = 2;
    }
  }
  const labelNames = ["sample", "edge", "izero"];
  if (nRegions > 3) {
    return { rowLabels, labelNames };
  }
  return { rowLabels, labelNames };
}

/**
 * Infer sample and izero bar positions from image profile.
 */
export function autoSampleIzeroRegions(
  image: number[][],
  qaxis: number[],
): [number, number, number, number] {
  const n = image.length;
  const profile = image.map((row) => row[row.length - 1] ?? 0);
  if (n < 4 || qaxis.length !== n) {
    const yMin = Math.min(...qaxis);
    const yMax = Math.max(...qaxis);
    const span = yMax - yMin;
    const margin = span * 0.05;
    return [yMin + span * 0.45, yMax - margin, yMin + margin, yMin + span * 0.35];
  }
  const kernel = [0.2, 0.2, 0.2, 0.2, 0.2];
  const smoothed = convolveSame(profile, kernel);
  let cliffIdx = 0;
  let maxGrad = 0;
  for (let i = 0; i < n - 1; i += 1) {
    const grad = Math.abs((smoothed[i + 1] ?? 0) - (smoothed[i] ?? 0));
    if (grad > maxGrad) {
      maxGrad = grad;
      cliffIdx = i;
    }
  }
  cliffIdx = Math.max(0, Math.min(n - 2, cliffIdx));
  const leftWinStart = Math.max(0, cliffIdx - 2);
  const leftWinEnd = cliffIdx + 1;
  const rightWinStart = cliffIdx + 1;
  const rightWinEnd = Math.min(n, cliffIdx + 4);
  let leftSum = 0;
  let leftCount = 0;
  for (let i = leftWinStart; i < leftWinEnd; i += 1) {
    leftSum += smoothed[i] ?? 0;
    leftCount += 1;
  }
  let rightSum = 0;
  let rightCount = 0;
  for (let i = rightWinStart; i < rightWinEnd; i += 1) {
    rightSum += smoothed[i] ?? 0;
    rightCount += 1;
  }
  const leftMean = leftCount > 0 ? leftSum / leftCount : 0;
  const rightMean = rightCount > 0 ? rightSum / rightCount : 0;
  const izeroOnLeft = leftMean < rightMean;
  const bufferPixels = Math.max(1, Math.floor(n * 0.05));
  const minRegionPixels = Math.max(2, Math.floor(n * 0.08));
  let barSampleLo: number;
  let barSampleHi: number;
  let barIzeroLo: number;
  let barIzeroHi: number;
  if (izeroOnLeft) {
    let izeroEnd = cliffIdx - bufferPixels;
    let sampleStart = cliffIdx + 1 + bufferPixels;
    izeroEnd = Math.max(minRegionPixels - 1, Math.min(n - 2, izeroEnd));
    sampleStart = Math.max(1, Math.min(n - minRegionPixels, sampleStart));
    if (izeroEnd < sampleStart) {
      barIzeroLo = qaxis[0] ?? 0;
      barIzeroHi = qaxis[izeroEnd] ?? 0;
      barSampleLo = qaxis[sampleStart] ?? 0;
      barSampleHi = qaxis[n - 1] ?? 0;
    } else {
      const mid = Math.floor(n / 2);
      barSampleLo = qaxis[0] ?? 0;
      barSampleHi = qaxis[Math.max(0, mid - minRegionPixels)] ?? 0;
      barIzeroLo = qaxis[Math.min(n - 1, mid + minRegionPixels)] ?? 0;
      barIzeroHi = qaxis[n - 1] ?? 0;
    }
  } else {
    let sampleEnd = cliffIdx - bufferPixels;
    let izeroStart = cliffIdx + 1 + bufferPixels;
    sampleEnd = Math.max(minRegionPixels - 1, Math.min(n - 2, sampleEnd));
    izeroStart = Math.max(1, Math.min(n - minRegionPixels, izeroStart));
    if (sampleEnd < izeroStart) {
      barSampleLo = qaxis[0] ?? 0;
      barSampleHi = qaxis[sampleEnd] ?? 0;
      barIzeroLo = qaxis[izeroStart] ?? 0;
      barIzeroHi = qaxis[n - 1] ?? 0;
    } else {
      const mid = Math.floor(n / 2);
      barIzeroLo = qaxis[0] ?? 0;
      barIzeroHi = qaxis[Math.max(0, mid - minRegionPixels)] ?? 0;
      barSampleLo = qaxis[Math.min(n - 1, mid + minRegionPixels)] ?? 0;
      barSampleHi = qaxis[n - 1] ?? 0;
    }
  }
  const span = Math.max(...qaxis) - Math.min(...qaxis);
  const margin = span * 0.02;
  if (Math.abs(barSampleHi - barSampleLo) < margin) {
    const lo = Math.min(barSampleLo, barSampleHi);
    const hi = Math.max(barSampleLo, barSampleHi);
    barSampleLo = lo - margin;
    barSampleHi = hi + margin;
  }
  if (Math.abs(barIzeroHi - barIzeroLo) < margin) {
    const lo = Math.min(barIzeroLo, barIzeroHi);
    const hi = Math.max(barIzeroLo, barIzeroHi);
    barIzeroLo = lo - margin;
    barIzeroHi = hi + margin;
  }
  return [barSampleLo, barSampleHi, barIzeroLo, barIzeroHi];
}

/**
 * Set dividing bar positions from 3-region segmentation (sample, edge, izero).
 */
export function barBoundsFromThreeRegions(
  image: number[][],
  qaxis: number[],
  profileColumns?: number,
  randomState = 0,
): [number, number, number, number] {
  const nRows = image.length;
  if (nRows < 3 || qaxis.length !== nRows) {
    const yMin = Math.min(...qaxis);
    const yMax = Math.max(...qaxis);
    const span = yMax - yMin;
    const margin = span * 0.05;
    return [yMin + span * 0.45, yMax - margin, yMin + margin, yMin + span * 0.35];
  }
  const { rowLabels } = segmentSpatialRegions(image, 3, profileColumns, randomState);
  const sampleRows: number[] = [];
  const izeroRows: number[] = [];
  for (let row = 0; row < nRows; row += 1) {
    if (rowLabels[row] === 0) {
      sampleRows.push(row);
    } else if (rowLabels[row] === 2) {
      izeroRows.push(row);
    }
  }
  if (sampleRows.length === 0 || izeroRows.length === 0) {
    return autoSampleIzeroRegions(image, qaxis);
  }
  const qSample = sampleRows.map((row) => qaxis[row] ?? 0);
  const qIzero = izeroRows.map((row) => qaxis[row] ?? 0);
  let barSampleLo = Math.min(...qSample);
  let barSampleHi = Math.max(...qSample);
  let barIzeroLo = Math.min(...qIzero);
  let barIzeroHi = Math.max(...qIzero);
  const span = Math.max(...qaxis) - Math.min(...qaxis);
  const margin = Math.max(span * 0.01, 1e-9);
  if (Math.abs(barSampleHi - barSampleLo) < margin) {
    const lo = Math.min(barSampleLo, barSampleHi);
    const hi = Math.max(barSampleLo, barSampleHi);
    barSampleLo = lo - margin;
    barSampleHi = hi + margin;
  }
  if (Math.abs(barIzeroHi - barIzeroLo) < margin) {
    const lo = Math.min(barIzeroLo, barIzeroHi);
    const hi = Math.max(barIzeroLo, barIzeroHi);
    barIzeroLo = lo - margin;
    barIzeroHi = hi + margin;
  }
  return [barSampleLo, barSampleHi, barIzeroLo, barIzeroHi];
}

export { sampleIzeroMasks } from "@/lib/stxm/sample-masks";

File: src/python/stxm/regions.py
import numpy as np
from sklearn.mixture import GaussianMixture


def segment_spatial_regions(
    image: np.ndarray,
    n_regions: int = 3,
    profile_columns: int | None = None,
    random_state: int = 0,
) -> tuple[np.ndarray, list[str]]:
    """
    Segment the spatial (row) axis into sample, edge, and izero (and optionally more sample sub-regions).
    Always finds exactly three base regions first: sample, edge, izero in spatial order
    (sample | edge | izero or izero | edge | sample). The edge is the thinnest of the three.
    For n_regions > 3, additional regions are found only inside the sample; edge and izero stay single.

    Parameters
    ----------
    image : np.ndarray
        2D scan (rows = spatial axis, cols = energy).
    n_regions : int
        Total regions: 3 = sample, edge, izero; 4+ = sample split into sample_1, sample_2, ... plus edge, izero.
    profile_columns : int, optional
        Number of trailing columns to average for the profile. If None, uses 20 or all columns.
    random_state : int
        Random state for GMM.

    Returns
    -------
    row_labels : np.ndarray
        Integer label per row. 0..n_regions-3 = sample parts, n_regions-2 = edge, n_regions-1 = izero.
    label_names : list of str
        Names: sample_1, sample_2, ..., edge, izero.
    """
    n_rows = image.shape[0]
    n_cols = image.shape[1]
    if profile_columns is None:
        profile_columns = min(20, n_cols)
    profile = np.mean(image[:, -profile_columns:], axis=1, dtype=np.float64).reshape(-1, 1)
    row_labels = np.zeros(n_rows, dtype=int)
    gm = GaussianMixture(n_components=3, random_state=random_state).fit(profile)
    raw = gm.predict(profile)
    extents = []
    means_intensity = []
    mean_row = []
    for idx in range(3):
        mask = raw == idx
        extents.append(np.sum(mask))
        means_intensity.append(float(np.mean(image[mask, :])))
        mean_row.append(float(np.mean(np.where(mask)[0])))
    edge_idx = int(np.argmin(extents))
    other = [i for i in range(3) if i != edge_idx]
    left_idx = other[0] if mean_row[other[0]] < mean_row[other[1]] else other[1]
    right_idx = other[1] if mean_row[other[0]] < mean_row[other[1]] else other[0]
    sample_idx = left_idx if means_intensity[left_idx] < means_intensity[right_idx] else right_idx
    izero_idx = right_idx if sample_idx == left_idx else left_idx
    row_labels[raw == sample_idx] = 0
    row_labels[raw == edge_idx] = 1
    row_labels[raw == izero_idx] = 2
    label_names = ["sample", "edge", "izero"]
    if n_regions > 3:
        n_sample_parts = n_regions - 2
        sample_mask = row_labels == 0
        n_sample = int(np.sum(sample_mask))
        if n_sample >= n_sample_parts * 2:
            profile_sample = profile[sample_mask].astype(np.float64)
            gm2 = GaussianMixture(n_components=n_sample_parts, random_state=random_state).fit(profile_sample)
            sub = gm2.predict(profile_sample)
            order = np.argsort(np.asarray(gm2.means_).ravel())
            remap = np.empty(n_sample_parts, dtype=int)
            remap[order] = np.arange(n_sample_parts)
            sub = remap[sub]
            row_labels[sample_mask] = sub
            row_labels[row_labels == 1] = n_regions - 2
            row_labels[row_labels == 2] = n_regions - 1
            label_names = [f"sample_{i+1}" for i in range(n_sample_parts)] + ["edge", "izero"]
    return row_labels, label_names


def bar_bounds_from_three_regions(
    image: np.ndarray,
    qaxis: np.ndarray,
    profile_columns: int | None = None,
    random_state: int = 0,
) -> tuple[float, float, float, float]:
    """
    Set the four dividing bar positions (sample_lo, sample_hi, izero_lo, izero_hi)
    from 3-region segmentation (sample, edge, izero). Uses segment_spatial_regions
    then maps sample and izero row indices to qaxis bounds.

    Parameters
    ----------
    image : np.ndarray
        2D scan (rows = spatial/qaxis axis, cols = energy).
    qaxis : np.ndarray
        Q-axis coordinates, length = image.shape[0].
    profile_columns : int, optional
        Passed to segment_spatial_regions.
    random_state : int
        Passed to segment_spatial_regions.

    Returns
    -------
    bar_sample_lo, bar_sample_hi, bar_izero_lo, bar_izero_hi : float
        Axis values for the four horizontal dividing bars.
    """
    qaxis = np.asarray(qaxis)
    n_rows = image.shape[0]
    if n_rows < 3 or len(qaxis) != n_rows:
        y_min, y_max = float(np.min(qaxis)), float(np.max(qaxis))
        span = y_max - y_min
        margin = span * 0.05
        return (
            y_min + span * 0.45,
            y_max - margin,
            y_min + margin,
            y_min + span * 0.35,
        )
    row_labels, _ = segment_spatial_regions(
        image,
        n_regions=3,
        profile_columns=profile_columns,
        random_state=random_state,
    )
    sample_rows = np.where(row_labels == 0)[0]
    izero_rows = np.where(row_labels == 2)[0]
    if sample_rows.size == 0 or izero_rows.size == 0:
        return auto_sample_izero_regions(image, qaxis)
    q_sample = qaxis[sample_rows]
    q_izero = qaxis[izero_rows]
    bar_sample_lo = float(np.min(q_sample))
    bar_sample_hi = float(np.max(q_sample))
    bar_izero_lo = float(np.min(q_izero))
    bar_izero_hi = float(np.max(q_izero))
    span = float(np.max(qaxis) - np.min(qaxis))
    margin = max(span * 0.01, 1e-9)
    if abs(bar_sample_hi - bar_sample_lo) < margin:
        lo, hi = min(bar_sample_lo, bar_sample_hi), max(bar_sample_lo, bar_sample_hi)
        bar_sample_lo = lo - margin
        bar_sample_hi = hi + margin
    if abs(bar_izero_hi - bar_izero_lo) < margin:
        lo, hi = min(bar_izero_lo, bar_izero_hi), max(bar_izero_lo, bar_izero_hi)
        bar_izero_lo = lo - margin
        bar_izero_hi = hi + margin
    return (bar_sample_lo, bar_sample_hi, bar_izero_lo, bar_izero_hi)


def sample_izero_masks(
    qaxis_points: np.ndarray,
    sample_lo: float,
    sample_hi: float,
    izero_lo: float,
    izero_hi: float,
) -> tuple[np.ndarray, np.ndarray]:
    """
    Boolean masks for sample and izero regions from axis coordinates.

    Parameters
    ----------
    qaxis_points : np.ndarray
        Q-axis (sample axis) coordinates, one per row.
    sample_lo, sample_hi : float
        Sample region bounds (inclusive).
    izero_lo, izero_hi : float
        Izero region bounds (inclusive).

    Returns
    -------
    sample_mask : np.ndarray
        True where qaxis_points is in [sample_lo, sample_hi].
    izero_mask : np.ndarray
        True where qaxis_points is in [izero_lo, izero_hi].
    """
    sample = (qaxis_points >= sample_lo) & (qaxis_points <= sample_hi)
    izero = (qaxis_points >= izero_lo) & (qaxis_points <= izero_hi)
    return sample, izero


def auto_sample_izero_regions(
    image: np.ndarray,
    qaxis: np.ndarray,
) -> tuple[float, float, float, float]:
    """
    Infer sample and izero bar positions from image profile (sample_lo, sample_hi, izero_lo, izero_hi).

    Parameters
    ----------
    image : np.ndarray
        2D scan (rows = qaxis).
    qaxis : np.ndarray
        Q-axis coordinates, length = image.shape[0].

    Returns
    -------
    bar_sample_lo, bar_sample_hi, bar_izero_lo, bar_izero_hi : float
        Axis values for the four horizontal bars.
    """
    profile = np.asarray(image[:, -1], dtype=float)
    n = len(profile)
    qaxis = np.asarray(qaxis)
    if n < 4 or len(qaxis) != n:
        y_min, y_max = float(np.min(qaxis)), float(np.max(qaxis))
        span = y_max - y_min
        margin = span * 0.05
        return (
            y_min + span * 0.45,
            y_max - margin,
            y_min + margin,
            y_min + span * 0.35,
        )
    kernel = np.ones(5) / 5.0
    profile = np.convolve(profile, kernel, mode="same")
    grad = np.diff(profile)
    cliff_idx = int(np.argmax(np.abs(grad)))
    cliff_idx = np.clip(cliff_idx, 0, n - 2)
    left_win = slice(max(0, cliff_idx - 2), cliff_idx + 1)
    right_win = slice(cliff_idx + 1, min(n, cliff_idx + 4))
    left_mean = float(np.mean(profile[left_win]))
    right_mean = float(np.mean(profile[right_win]))
    izero_on_left = left_mean < right_mean
    buffer_pixels = max(1, int(n * 0.05))
    min_region_pixels = max(2, int(n * 0.08))
    if izero_on_left:
        izero_end = cliff_idx - buffer_pixels
        sample_start = cliff_idx + 1 + buffer_pixels
        izero_end = np.clip(izero_end, min_region_pixels - 1, n - 2)
        sample_start = np.clip(sample_start, 1, n - min_region_pixels)
        if izero_end < sample_start:
            bar_izero_lo = float(qaxis[0])
            bar_izero_hi = float(qaxis[izero_end])
            bar_sample_lo = float(qaxis[sample_start])
            bar_sample_hi = float(qaxis[n - 1])
        else:
            mid = n // 2
            bar_sample_lo = float(qaxis[0])
            bar_sample_hi = float(qaxis[max(0, mid - min_region_pixels)])
            bar_izero_lo = float(qaxis[min(n - 1, mid + min_region_pixels)])
            bar_izero_hi = float(qaxis[n - 1])
    else:
        sample_end = cliff_idx - buffer_pixels
        izero_start = cliff_idx + 1 + buffer_pixels
        sample_end = np.clip(sample_end, min_region_pixels - 1, n - 2)
        izero_start = np.clip(izero_start, 1, n - min_region_pixels)
        if sample_end < izero_start:
            bar_sample_lo = float(qaxis[0])
            bar_sample_hi = float(qaxis[sample_end])
            bar_izero_lo = float(qaxis[izero_start])
            bar_izero_hi = float(qaxis[n - 1])
        else:
            mid = n // 2
            bar_izero_lo = float(qaxis[0])
            bar_izero_hi = float(qaxis[max(0, mid - min_region_pixels)])
            bar_sample_lo = float(qaxis[min(n - 1, mid + min_region_pixels)])
            bar_sample_hi = float(qaxis[n - 1])
    span = float(np.max(qaxis) - np.min(qaxis))
    margin = span * 0.02
    if abs(bar_sample_hi - bar_sample_lo) < margin:
        lo, hi = min(bar_sample_lo, bar_sample_hi), max(bar_sample_lo, bar_sample_hi)
        bar_sample_lo = lo - margin
        bar_sample_hi = hi + margin
    if abs(bar_izero_hi - bar_izero_lo) < margin:
        lo, hi = min(bar_izero_lo, bar_izero_hi), max(bar_izero_lo, bar_izero_hi)
        bar_izero_lo = lo - margin
        bar_izero_hi = hi + margin
    return (bar_sample_lo, bar_sample_hi, bar_izero_lo, bar_izero_hi)
